# Alunos

Matheus Dos Santos Palheta -22052572 <br>
Matheus Silva dos Santos - 22052573 <br>
Vinícius Luiz Nunes da Fonseca - 22050031

## Tarefa 1

### Saída de instalação do PostgreSQL

A instalação do PostgreSQL foi feita diretamente do site oficial, inclusive dentro do container docker criado! <br>
Foi gerado o seguinte log no terminal após a instalação:

In [ ]:
!pg_config --version

## Tarefa 2

### Fase de preparação

In [ ]:
%cd /app/datadir/tpch/

%run tpch_pgsql.py prepare

### Fase de carregamento

In [ ]:
%cd /app/datadir/tpch/

%run tpch_pgsql.py load

## Tarefa 3

### Execução das consultas

In [ ]:
%cd /app/datadir/tpch/

%run tpch_pgsql.py query 

## Tarefa 4

### Informações Sistema Operacional:

PRETTY_NAME="Ubuntu 22.04.2 LTS" <br>
VERSION="22.04.2 LTS (Jammy Jellyfish)"<br>
ID=ubuntu<br>
ID_LIKE=debian<br>
HOME_URL="https://www.ubuntu.com/"<br>
SUPPORT_URL="https://help.ubuntu.com/"<br>
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"<br>
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"<br>
UBUNTU_CODENAME=jammy<br>

### informações CPU
Architecture:            x86_64 <br>
  Address sizes:         48 bits physical, 48 bits virtual<br>
  Byte Order:            Little Endian<br>
Vendor ID:               AuthenticAMD<br>
  Model name:            AMD Ryzen 5 5600G with Radeon Graphics<br>
    CPU family:          25<br>
    Model:               80<br>
    Thread(s) per core:  2<br>
    Core(s) per socket:  6<br>
    Socket(s):           1<br>
    Stepping:            0<br>
    BogoMIPS:            7785.09<br>

### Informação memória

      total        used        free      shared  buff/cache   available
      6.7Gi       671Mi       5.2Gi        16Mi       852Mi       5.8Gi mem
      2.0Gi          0B       2.0Gi                                     swap
      
      Filesystem      Size  Used Avail Use% Mounted on
      drvfs           475G  338G  138G  72% /mnt/c

## Tarefa 5

**Resposta:** A Tarefa 5 não pôde ser respondida no Jupyter Notebook porque os computadores utilizados possuem apenas unidades de estado sólido (SSDs) e não possuem discos rígidos (HDs). Os parâmetros solicitados na tarefa são específicos dos discos rígidos e não se aplicam aos SSDs. Portanto, não foi possível coletar os dados necessários para responder à tarefa devido às diferenças técnicas entre os dois tipos de armazenamento.

## Tarefa 6

| Característica                                 | Descrição                                                                                                                                                                                                                          |
|-----------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Modelo de Armazenamento                        | O PostgreSQL utiliza o modelo de armazenamento relacional, baseado em tabelas, colunas e linhas, seguindo os princípios do modelo ACID (Atomicidade, Consistência, Isolamento e Durabilidade).                                      |
| Formato de Armazenamento                       | O PostgreSQL armazena os dados em páginas de tamanho fixo (geralmente 8 KB por página), onde cada página pode conter várias tuplas. Os dados são organizados em segmentos, tablespaces e bancos de dados.                          |
| Controle de Acesso aos Dados                   | O PostgreSQL oferece um sistema de controle de acesso robusto e flexível, permitindo a definição de permissões de acesso a nível de objeto (tabelas, colunas, etc.) e a nível de usuário/grupo.                                     |
| Indexação de Dados                             | O PostgreSQL suporta vários tipos de índices, incluindo índices B-tree, hash, GIN (Generalized Inverted Index) e GiST (Generalized Search Tree), permitindo a otimização do acesso aos dados em diferentes cenários.               |
| Compressão de Dados                            | O PostgreSQL oferece suporte à compressão de dados por meio de técnicas como TOAST (The Oversized-Attribute Storage Technique) e tabelas particionadas, permitindo economia de espaço de armazenamento.                             |
| Gerenciamento de Espaço em Disco               | O PostgreSQL utiliza um sistema de gerenciamento de espaço em disco que aloca automaticamente espaço para novos dados e realiza reutilização de espaço para dados removidos ou atualizados.                                         |
| Recuperação de Falhas (Recovery)               | O PostgreSQL possui um mecanismo de recuperação robusto, com suporte a logs de transação (WAL - Write-Ahead Logging) que permite a recuperação do banco de dados para um estado consistente após uma falha.                         |
| Replicação de Dados                            | O PostgreSQL oferece várias opções de replicação de dados, incluindo replicação baseada em registros (Streaming Replication), replicação síncrona e assíncrona, permitindo alta disponibilidade e escalabilidade.                      |
| Particionamento de Tabelas                     | O PostgreSQL suporta o particionamento de tabelas, permitindo que grandes conjuntos de dados sejam divididos em partições menores com base em um critério especificado, o que facilita o gerenciamento de grandes volumes de dados. |
| Recursos Avançados de Consulta e Otimização    | O PostgreSQL possui um otimizador de consultas avançado que analisa as consultas submetidas e escolhe o plano de execução mais eficiente. Além disso, oferece recursos como estatísticas de tabelas, views materializadas e otimização de expressões.  |


## Tarefa 7

| Sistema de Arquivo | Descrição | Capacidade Máxima de Arquivo | Capacidade Máxima de Sistema de Arquivo | Journaling | Fragmentação | Suporte a ACLs | Compressão | Transações |
|--------------------|-----------|------------------------------|---------------------------------------|------------|--------------|----------------|------------|------------|
| Ext2               | O Ext2 é um sistema de arquivos de primeira geração para o Linux. | 2 TB | 4 TB | Não | Propenso à fragmentação | Não | Não | Não |
| Ext3               | O Ext3 é uma extensão do Ext2 que adiciona journaling para maior segurança. | 2 TB | 4 TB | Sim | Propenso à fragmentação | Não | Não | Não |
| ReiserFS           | O ReiserFS é um sistema de arquivos de alto desempenho, conhecido por sua rápida manipulação de pequenos arquivos. | 8 TB | 16 TB | Sim | Melhor desempenho com arquivos pequenos | Sim | Não | Não |
| XFS                | O XFS é um sistema de arquivos de alto desempenho, otimizado para escalabilidade e grandes sistemas de armazenamento. | 8 EB | 8 EB | Sim | Melhor desempenho com arquivos grandes | Sim | Sim | Sim |

Observações:

    Capacidade máxima de arquivo refere-se ao tamanho máximo de um único arquivo suportado pelo sistema de arquivos.
    Capacidade máxima de sistema de arquivos refere-se ao tamanho máximo total do sistema de arquivos.
    Journaling refere-se à técnica de registro de transações para garantir a consistência dos dados após falhas de energia ou reinicializações inesperadas.
    Fragmentação indica a tendência do sistema de arquivos em fragmentar os arquivos em várias partes físicas do disco.
    ACLs (Access Control Lists) permitem a atribuição de permissões mais granulares a arquivos e diretórios.
    Compressão indica se o sistema de arquivos suporta compressão de dados para economizar espaço de armazenamento.
    Transações referem-se à capacidade do sistema de arquivos de executar operações atômicas (tudo ou nada).

## Tarefa 8

In [1]:
%cd /app/datadir/scripts/tarefa8

/app/datadir/scripts/tarefa8


In [7]:
%run create_tables.py

Tabelas criadas com sucesso!


In [3]:
%run arquivo1.py

Arquivo 1 gerado com sucesso!

Arquivo 1 inserido com sucesso no banco de dados!

Arquivo 1 excluído com sucesso! (Para liberar espaço no disco)

Tempo para criação do arquivo: 1.42 segundos
Tempo de inserção no banco de dados: 25.64 segundos
Tempo de execução total: 27.06 segundos
Espaço ocupado no disco pelo arquivo: 563200000 bytes

Tabela excluída com sucesso! (Para liberar espaço no banco de dados)


In [8]:
%run arquivo2.py

Arquivo 2 gerado com sucesso!

Arquivo 2 inserido com sucesso no banco de dados!

Arquivo 2 excluído com sucesso! (Para liberar espaço no disco)

Tempo para criação do arquivo: 0.32 segundos
Tempo de inserção no banco de dados: 2.7 segundos
Tempo de execução total: 3.02 segundos
Espaço ocupado no disco pelo arquivo: 92160000 bytes

Tabela excluída com sucesso! (Para liberar espaço no banco de dados)


In [5]:
%run arquivo3.py

Arquivo 3 gerado com sucesso!

Arquivo 3 inserido com sucesso no banco de dados!

Arquivo 3 excluído com sucesso! (Para liberar espaço no disco)

Tempo para criação do arquivo: 8.57 segundos
Tempo de inserção no banco de dados: 145.36 segundos
Tempo de execução total: 153.92 segundos
Espaço ocupado no disco pelo arquivo: 3379200000 bytes

Tabela excluída com sucesso! (Para liberar espaço no banco de dados)


In [6]:
%run arquivo4.py

Arquivo 4 gerado com sucesso!

Arquivo 4 inserido com sucesso no banco de dados!

Arquivo 4 excluído com sucesso! (Para liberar espaço no disco)

Tempo para criação do arquivo: 237.62 segundos
Tempo de inserção no banco de dados: 20.85 segundos
Tempo de execução total: 258.47 segundos
Espaço ocupado no disco pelo arquivo: 460645710 bytes

Tabela excluída com sucesso! (Para liberar espaço no banco de dados)


## Tarefa 9

## Tarefa 10

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run cria_tabela.py

## Tarefa 11

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run numero_paginas.py

## Tarefa 12

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run numero_blocos.py

## Tarefa 13

### Qual o tempo gasto para realizar uma consulta para um valor (tendo a tabela 100000 tuplas)?
### Qual o tempo gasto para recriar um índice para o atributo ‘v’?


In [ ]:
%cd /app/datadir/scripts/tabela_t

%run tempo_100000.py

### Remova a tabela ‘t’ e crie novamente com 1.0000.000 de tuplas
### Qual o tempo gasto para realizar uma consulta para um valor específico?
### Qual o tempo gasto para recriar um índice para o atributo ‘v’?


In [ ]:
%cd /app/datadir/scripts/tabela_t

%run tempo_1000000.py

## Tarefa 14

### 100.000 tuplas

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run fill_factor_100000.py

### 1.000.000 tuplas

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run fill_factor_1000000.py

## Tarefa 15

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run cria_tabela_movie.py

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run cria_tabela_actor.py

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run cria_tabela_casting.py

## Tarefa 16

### repetindo tarefas 11, 12 e 13 com indice decrescente para 100.000 tuplas

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run indice_decrescente_100000.py

### repetindo tarefas 11, 12 e 13 com indice decrescente para 1.000.000 tuplas

In [ ]:
%cd /app/datadir/scripts/tabela_t

%run indice_decrescente_1000000.py

## Tarefa 17

### Letra A

Execute o script movie.sql para criar as tabelas e índices e carregar os dados necessários às próximas atividades

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run cria_tabela_movie.py

### Letra B

Verifique no catálogo do banco de dados os seguintes metadados sobre os índices associados às tabelas e apresente-os no relatório: Nome do índice, nome da tabela, altura, número máximo de chaves por bloco, número médio de chaves por bloco, número de blocos folha, número de médio de blocos folha por chave, número médio de blocos de dados por chave, número de linhas e número de chaves distintas. 

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run metadados_tarefa17.py

## Tarefa 18

### Letra A

Consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número pequeno de tuplas (<10 tuplas). Em seguida, é executado o comando EXPLAIN sobre esta consulta e apresentado os resultados.

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa18.1.py

### Letra B

Consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número grande de tuplas (>80% das tuplas). Execute o comando EXPLAIN sobre esta consulta e apresente os resultados.

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa18.2.py

### Letra C

Explique porque o índice sobre VOTES não é sempre usado nas consultas sobre este atributo.

**Resposta:** O uso de um índice em consultas sobre o atributo "votes" pode não ser sempre vantajoso devido a diferentes fatores, como a distribuição dos dados na coluna indexada, o tamanho da tabela e o valor de limite (LIMIT) especificado na consulta. O otimizador de consulta do PostgreSQL decide se utilizar um índice ou uma varredura sequencial com base na estimativa de custo de cada abordagem. Se a distribuição dos dados é uniforme, a tabela é pequena ou o limite é grande, pode ser mais eficiente fazer uma varredura sequencial em todos os registros em vez de usar um índice. O PostgreSQL avalia esses fatores para determinar a estratégia de execução mais eficiente em cada consulta específica.

No primeiro EXPLAIN, a consulta utiliza o índice **movie_votes** na busca, conforme indicado pela linha **'Index Scan using movie_votes on movie'**. Isso significa que o otimizador de consulta do PostgreSQL decidiu utilizar o índice para melhorar a eficiência da busca dos registros que atendem à condição **'votes > 10000'**. O custo da operação de busca usando o índice foi estimado como **'cost=0.28..10.33'**.

No segundo EXPLAIN, a consulta realiza uma varredura sequencial (Seq Scan) na tabela **'movie'**, conforme indicado pela linha **'Seq Scan on movie'**. Nesse caso, não está sendo utilizado um índice para a busca. A condição **'votes > 1'** é aplicada a todos os registros durante a varredura sequencial. O custo da varredura sequencial foi estimado como **'cost=0.00..38.05'**.

## Tarefa 19

### Letra A

Resultado do comando explain sobre as duas consultas abaixo:

SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie); <br>
SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa19.py

### Letra B

Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique?

**Resposta:** A primeira consulta utiliza um "Index Scan" e é mais eficiente do que a segunda consulta, que utiliza um "Seq Scan". O "Index Scan" aproveita um índice existente na coluna "votes" para acessar diretamente as linhas que atendem à condição da consulta, enquanto o "Seq Scan" faz uma verificação sequencial em toda a tabela. Portanto, em termos de eficiência, a primeira consulta é mais recomendada. No entanto, a eficiência de uma consulta depende de vários fatores e é importante avaliar o desempenho real no contexto específico.

## Tarefa 20

Considere as duas consultas equivalentes em SQL a seguir, as quais retornam os filmes com mais votos que “Star Wars” 

SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars'); 

SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';

### Letra A

Apresente o resultado do comando explain sobre as duas consultas acima

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa20.py

### Letra B

Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique?

Sim, existem diferencas entre as consultas. Na primeira, eh utilizado um indice para filtar os resultados e a condicao do indice eh aplicada diretamente, resultando num acesso mais eficiente dos dados com o custo entre 8.57 e 43.34. Na segunda, o plano de execução, tendo um custo estimado entre 0.56 e 49.49, utiliza "Nested Loop" que eh menos eficiente em comparação com o uso direto de um índice, alem de "m1.votes > m2.vote" envolver a comparação de duas colunas, o que pode exigir mais recursos de processamento. Em resumo, o primeiro plano de consulta apresentado tem uma estimativa de melhor desempenho quando comparado ao segundo plano de consulta.

## Tarefa 21

Considere as seguintes consultas SQL sobre o atributo  TITLE usando o operador LIKE. 

SELECT title FROM movie WHERE title LIKE 'I%'; 

SELECT title FROM movie WHERE substr(title, 1, 1) = 'I'; 

SELECT title FROM movie WHERE title LIKE '%A'; 

### Letra A

Apresente o resultado do comando explain sobre as três consultas acima

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa21.py

### Letra B

Qual das três apresenta o menor custo? Porque?

**Resposta:** A consulta 1 eh a de menor custo. A operação LIKE é utilizada para filtrar as linhas da tabela movie onde o título começa com a letra 'I', tendo um custo de 0.00..38.05. O custo eh o menor porque a tabela eh percorrida sequencialmente (Seq Scan) e o filtro eh aplicado diretamente no campo title usando a operação. Em resumo, a primeira eh de menor custo pq aplica o LIKE diretamente no campo 'title'.

### Letra C
O índice sobre TITLE foi usado para todas elas? Justifique. 

## Tarefa 22 

Verificação da hipótese de distribuição uniforme na estimativa de seletividade

Considere as seguintes  consultas sobre o atributo TITLE da tabela MOVIE

SELECT title FROM movie WHERE votes < 1000; 

SELECT title FROM movie WHERE votes > 40000;


### Letra A
Apresente o resultado do comando explain sobre as duas consultas acima. Explique o resultado.

In [ ]:
%cd /app/datadir/scripts/movie_catalog/

%run explain_tarefa22.py

**Resposta:** As duas utilizam um indice para otimizar a consulta, sendo ele "movie_votes". Na primeira consulta, o custo eh de 20.02 unidades, com um custo inicial de 0.28 para acessar o indice e o resto do custo eh para acessar as linhas de movie. Na segunda consulta, o custo total eh de 8.42 com um custo inicial de 0.28 para acessar o indice e o resto para acessar as linhas de movie. A diferenca entre os custos da-se pela forma como as consultas foram realizadas: a primeira busca "menores que" e a segunda busca "maiores que". As buscas utilizando ">" sao mais eficientes porque os intervalos na arvore sao ordenados de forma crescente, permitindo que a consulta 2 seja mais eficiente.

### Letra B
Compare o número de tuplas selecionadas por cada consulta. Qual das duas tem a menor seletividade?

**Resposta:** 

Primeira consulta: 328 linhas.
Segunda consulta: 8 linhas. 

Com base nesses estimatias, a Consulta 2 tem menor seletividade porque espera-se que retorne menos linhas em comparacao à Consulta 1.

# PARTE V

## Tarefa 23

## Tarefa 24

## Tarefa 25

## Tarefa 26